test/tests/constraints/equal_value_embedded_constraint/tests

```diff
@@ -119,8 +110,95 @@
+      exodiff = '1D_2D_ADpenalty_out.e'
+      abs_zero = 1e-9
+      mesh_mode = REPLICATED
+      detail = "1D and 2D;"
+    []
+    [1D_3D]
+      type = 'Exodiff'
+      input = 'embedded_constraint.i'
+      cli_args = 'Mesh/file=gold/1D_3D.e Constraints/equal/formulation=penalty '
+                 'Constraints/equal/type=ADEqualValueEmbeddedConstraint Outputs/file_base=1D_3D_ADpenalty_out'
+      exodiff = '1D_3D_ADpenalty_out.e'
+      abs_zero = 1e-9
+      mesh_mode = REPLICATED
+      detail = "1D and 3D;"
+    []
+    [2D_2D]
+      type = 'Exodiff'
+      input = 'embedded_constraint.i'
+      cli_args = 'Mesh/file=gold/2D_2D.e Constraints/equal/formulation=penalty '
+                 'Constraints/equal/type=ADEqualValueEmbeddedConstraint Outputs/file_base=2D_2D_ADpenalty_out'
+      exodiff = '2D_2D_ADpenalty_out.e'
+      abs_zero = 1e-9
+      mesh_mode = REPLICATED
+      detail = "2D;"
+    []
+    [2D_3D]
+      type = 'Exodiff'
+      input = 'embedded_constraint.i'
+      cli_args = 'Mesh/file=gold/2D_3D.e Constraints/equal/formulation=penalty '
+                 'Constraints/equal/type=ADEqualValueEmbeddedConstraint Outputs/file_base=2D_3D_ADpenalty_out'
+      exodiff = '2D_3D_ADpenalty_out.e'
+      abs_zero = 1e-9
+      mesh_mode = REPLICATED
+      detail = "2D and 3D; and"
+    []
+    [3D_3D]
+      type = 'Exodiff'
+      input = 'embedded_constraint.i'
+      cli_args = 'Mesh/file=gold/3D_3D.e Constraints/equal/formulation=penalty '
+                 'Constraints/equal/type=ADEqualValueEmbeddedConstraint Outputs/file_base=3D_3D_ADpenalty_out'
+      exodiff = '3D_3D_ADpenalty_out.e'
+      abs_zero = 1e-9
+      mesh_mode = REPLICATED
       detail = "3D subdomains."
     []
   []
+  [error_kinematic_AD]
+    design = 'source/constraints/ADEqualValueEmbeddedConstraint.md'
+    requirement = "The system shall error when kinematic constraints are used with automatic differentiation."
+    issues = '#30499'
+    type = RunException
+    input = 'embedded_constraint.i'
+    cli_args = 'Mesh/file=gold/1D_2D.e Constraints/equal/formulation=kinematic '
+               'Constraints/equal/type=ADEqualValueEmbeddedConstraint Executioner/solve_type=NEWTON'
+    expect_err = "AD constraints cannot be used with KINEMATIC formulation."
+    mesh_mode = REPLICATED
+  []
+  [two_var_AD]
+    design = 'source/constraints/ADEqualValueEmbeddedConstraint.md'
+    requirement = "The system shall couple two different variables using automatic "
+                  "differentiation and penalty constraint and it will produce the same "
+                  "results as a single variable simulation."
+    issues = '#30499'
+    type = RunApp
+    input = 'two_var_embedded_constraint.i'
+    mesh_mode = REPLICATED
+  []
+  [error_two_var_AD]
+    design = 'source/constraints/ADEqualValueEmbeddedConstraint.md'
+    requirement = "The system shall error when the nonAD constraint is used for coupling two variables."
```

Comment:
does it converge with PJFNK ? is the Jacobian is not 100% correct witohut AD, makes sense that it does not converge with Newton

Reply:
I didn't try PJFNK.  The Jacobian is wrong but I didn't look at how wrong.  The residual diverges.

Reply:
Let's try it. These constraints were meant to be used with PJFNK since the Jacobians were not correct

Reply:
We had assumed the nonAD version of these constraints had a correct Jacobian.  We didn't know it was incorrect until I added the test in this PR that coupled two different variables.  The nonAD version of this constraint works with Newton when this constraint is coupling the same variable.

Reply:
>  It does not work for kinematic and PJFNK

It worked fine on the test case. you just need to adjust the scaling.
But you're right, the jacobian is terribly wrong. 

We need an issue for this. 
I m fine with the error

Reply:
You are right, I purposefully put in bad manual scaling to highlight the issue.  It would converge without the bad scaling.  I added an issue that links to this PR and test.

Reply:
ok found one error which prevents getting the correct Jacobian
for off-diag on primary side, we are actually using the wrong dofs (sending to secondary side)
should be
```

                  // Cache the jacobian block for the primary side
                  nec->addJacobian(_fe_problem.assembly(0, number()),
                                   nec->_Kne,
                                   nec->primaryVariable().dofIndicesNeighbor(),
                                   nec->_connected_dof_indices,
                                   nec->primaryVariable().scalingFactor());
```

but it s not the only problem.
the getConnectedDofs is not finding any connected DOFs for the primary variables

Reply:
how did you find that, are stepping through this in gdb?

Reply:
the node_to_elem maps, is not going to get elements that live on the primary side? 
I think we need more logic in getConnectedDofs to go grab those

Reply:
why is this test problem so big 😦